wdl-analysis/src/queue.rs

```diff
@@ -239,6 +259,57 @@ where
                         graph.get_mut(node).notify_change(discard_pending);
                     }
                 }
+                Request::Format(FormatRequest {
+                    document,
+                    completed,
+                }) => {
+                    let graph = self.graph.read();
+
+                    let result = graph
+                        .get_index(&document)
+                        .and_then(|index| {
+                            graph.get(index).document().and_then(|document| {
+                                match graph.get(index).parse_state() {
+                                    // NOTE: if we haven't parsed the document yet, then
+                                    // we don't have the line lengths of the document,
+                                    // so we can't proceed with formatting and we should
+                                    // just silently return.
+                                    ParseState::NotParsed | ParseState::Error(_) => None,
+                                    ParseState::Parsed {
+                                        version: _,
+                                        root: _,
+                                        lines,
+                                        diagnostics,
```

Comment:
This is just "parse" diagnostics, correct? The output of running `wdl_ast::Document::parse()`? Just checking my understanding that this does not include analysis diagnostics.

Reply:
Well, this does include _validation_ diagnostics. If you go look at the definition of the struct, it's a `wdl_analysis::graph::Diagnostics`.

That being said, we want to catch errors whether they happen during parsing or validation and decline to format in those cases. For example, if you include two input blocks, that fails _validation_, but not parsing. In that case, if you don't catch the validation errors, it will just delete one of the input blocks (because the assumption that the WDL has been validated is inherent in the formatting code).

There might even be a case to be made that we don't depend on the caller to ensure that the WDL is validated when calling format—instead, we require programmatically that the WDL parses and validates.

Reply:
Correct me if wrong, but as I understand there are 3 classes of errors we have to deal with:
- **parse** errors (like a missing close delimiter)
- **validation** errors (like two input blocks for the same task)
- **analysis** errors (mismatched type or _any class of error in an import_)

Formatting a document with either parse or validation errors will lead to odd behavior (such as things getting deleted). But I think we should tolerate analysis errors, as none of the format code makes assumptions about analysis validity.

A bad import shouldn't prohibit formatting IMO.

Reply:
Yes—what you describe is my understanding of how the current code will work.

Reply:
That's correct, these are either the parse diagnostics or the validation diagnostics (if there were no parse diagnostics), so this is the correct collection to check to see if we should format the document.

Reply:
Sounds good. Gets my 👍